src/amsterdam_schema/publish/cli.py

```diff
@@ -86,15 +86,43 @@ def fetch_local_as_publishable(
     return publishable_paths
 
 
-def get_index_file_obj(publishable_paths: List[List[str]]) -> BytesIO:
-    index = {}
-    for path_parts in publishable_paths:
+def get_index_file_obj(publishable_paths: List[List[str]], files_root: Path) -> BytesIO:
+    """Create index file of all the datasets in JSON format.
+
+    For every dataset.json file it will add an entry of the form:
+    `dataset_id:'path/to/dataset'`.
+    When a versions.json file is encountered, its path will be included
+    and any dataset.json files with the same id in subdirectories ignored.
+    """
+    index: Dict[str, str] = {}
+    versions_file = "versions.json"
+    dataset_file = "dataset.json"
```

Comment:
Both  `versions_file` and `dataset_file` seem to be local constants. If they are maybe you could consider renaming them, per PEP 008, to `VERSIONS_FILE` and `DATASET_FILE` respectively. In addition you might want to annotate them with `: Final`. That way MyPy checks whether they are actually used as constants.

```python
from typing import Final
...
VERSIONS_FILE: Final = "versions.json"
DATASET_FILE: Final = "dataset.json"
```

Reply:
They should be all caps. Have to disagree on the typing though. Just extra verbosity for protection against something that is clearly wrong from context, would cause obvious failures anyway and pylint already warns you about. Falls on the wrong side of the nuisance/benefit scale for me.

Reply:
Except we do not use `pylint`. We do use `flake8` and that does not detect it:

```python
def fubar():
    A_CONSTANT = 42
    A_CONSTANT = 299_792_458  # speed of light in m/s
    print(A_CONSTANT)


fubar()
```
```
flake8 scratch_61.py 
scratch_61.py:1:1: D103 Missing docstring in public function
scratch_61.py:4:5: T001 print found.
```
With type annotations this becomes;
```python
from typing import Final


def fubar() -> None:
    A_CONSTANT: Final = 42
    A_CONSTANT: Final = 299_792_458  # speed of light in m/s
    print(A_CONSTANT)


fubar()
```
and MyPy complains rightfully so:
```
mypy scratch_61.py
scratch_61.py:6:5: error: Cannot redefine an existing name as final  [misc]
Found 1 error in 1 file (checked 1 source file)
```
Anyway I've made my case. Final (pun intended) decision is yours as we don't have a formal rule that mandates type annotations in every possible case.